### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [3]:
Total_players=pd.DataFrame({"Total Players":[len(purchase_data["SN"].unique())]})
Total_players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
Summary_df=pd.DataFrame({
    "Number of Unique Items":[len(purchase_data["Item Name"].unique())],
    "Average Price":['${:.2f}'.format(purchase_data["Price"].mean())],
    "Number of Purchases":[len(purchase_data["Purchase ID"].unique())],
    "Total Revenue":['${:.2f}'.format(purchase_data["Price"].sum())]
})
Summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780      $2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [5]:
 def percentage(x):
      pct = float(x/SN_Clean["Gender"].count() * 100)
      return round(pct, 2)

In [6]:

SN_Clean=purchase_data.drop_duplicates(subset=['SN'])

Gender_Demographics=SN_Clean["Gender"].value_counts()

Gender_percen=['{:.2f}%'.format(x) for x in (Gender_Demographics.apply(percentage))]

GD_df=pd.DataFrame({
    "Total Count":Gender_Demographics,
    "Percentage of Players":Gender_percen
})
GD_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
p_a=purchase_data.groupby("Gender")
P_count=p_a["Price"].count()

In [8]:
Av_pc=['${:.2f}'.format(x) for x in p_a["Price"].mean()]

In [9]:
T_pv=['${:.2f}'.format(x) for x in p_a["Price"].sum()]

In [10]:
s_a=SN_Clean.groupby("Gender")
Av_tpc=['${:.2f}'.format(x) for x in s_a["Price"].mean()]

In [11]:
pag=pd.DataFrame({
    "Purchase Count":P_count,
    "Average Purchase Price":Av_pc,
    "Total Purchase Value":T_pv,
    "Avg Total Purchase per Person":Av_tpc
})
pag

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $3.17  
Male                              $1967.64                         $3.05  
Other / Non-Disclosed               $50.19                         $3.41

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [12]:
bins = [0,10, 15, 19, 25, 30,35,39,50]
group_names=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

In [13]:
SN_Clean["Age Demographic"] = pd.cut(SN_Clean["Age"], bins, labels=group_names, include_lowest=False)
SN_Clean

<ipython-input-13-dd9dc053704e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  SN_Clean["Age Demographic"] = pd.cut(SN_Clean["Age"], bins, labels=group_names, include_lowest=False)


Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
773          773         Hala31   21    Male       19   
774          774     Jiskjask80   11    Male       92   
775          775     Aethedru70   21  Female       60   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male       92   

                                     Item Name  Price Age Demographic  
0    Extraction, Quickblade Of Trembling Hands   3.53           20-24  
1                            Frenzied Scimitar   1.56             40+  
2                                 Final Critic   4.88           20-24  
3                                  Blindscythe   3.27           20-24  
4                                         Fury   1.44           20-24  
..                                         ...    ...             ...  
773              Pursuit, Cudgel of Necromancy   1.02           20-24  
774                               Final Critic   4.19           10-14  
775                                       Wolf   3.54           20-24  
777      Celeste, Incarnation of the Corrupted   3.46           20-24  
778                               Final Critic   4.19             <10  

[576 rows x 8 columns]

In [14]:
 def percent(x):
      pct = float(x/SN_Clean["Age Demographic"].count() * 100)
      return round(pct, 2)

In [15]:
A_D=SN_Clean.groupby("Age Demographic")
T_C=A_D["SN"].count()
T_CP=['{:.2f}%'.format(x) for x in T_C.apply(percent)]
T_Cdf=pd.DataFrame({
    "Total Count":T_C,
    "Percentage of Players":T_CP
})
T_Cdf

Total Count Percentage of Players
Age Demographic                                   
<10                       24                 4.17%
10-14                     41                 7.12%
15-19                     81                14.06%
20-24                    301                52.26%
25-29                     59                10.24%
30-34                     37                 6.42%
35-39                     21                 3.65%
40+                       12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [16]:
bins = [0,10, 15, 19, 25, 30,35,39,50]
group_names=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

In [17]:
purchase_data["Age Demographic"] = pd.cut(purchase_data["Age"], bins, labels=group_names, include_lowest=False)
purchase_data

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male       92   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Demographic  
0    Extraction, Quickblade Of Trembling Hands   3.53           20-24  
1                            Frenzied Scimitar   1.56             40+  
2                                 Final Critic   4.88           20-24  
3                                  Blindscythe   3.27           20-24  
4                                         Fury   1.44           20-24  
..                                         ...    ...             ...  
775                                       Wolf   3.54           20-24  
776                           Exiled Doomblade   1.63           20-24  
777      Celeste, Incarnation of the Corrupted   3.46           20-24  
778                               Final Critic   4.19             <10  
779                                       Dawn   4.60           20-24  

[780 rows x 8 columns]

In [18]:
PA_AD=purchase_data.groupby("Age Demographic")
Pur_c=PA_AD["Price"].count()
Pur_c

Age Demographic
<10       32
10-14     54
15-19    101
20-24    424
25-29     77
30-34     52
35-39     27
40+       13
Name: Price, dtype: int64

In [19]:
Pur_av=['${:.2f}'.format(x) for x in PA_AD["Price"].mean()]
Pur_av

['$3.40', '$2.90', '$3.04', '$3.06', '$2.88', '$2.99', '$3.54', '$2.94']

In [20]:
Pur_sum=['${:.2f}'.format(x) for x in PA_AD["Price"].sum()]
Pur_sum

['$108.96',
 '$156.60',
 '$307.24',
 '$1295.96',
 '$221.42',
 '$155.71',
 '$95.64',
 '$38.24']

In [21]:
A_D=SN_Clean.groupby("Age Demographic")
Pur_C=['${:.2f}'.format(x) for x in A_D["Price"].mean()]
Pur_C

['$3.42', '$2.94', '$3.14', '$3.06', '$2.90', '$3.00', '$3.36', '$3.04']

In [22]:
Pur_df=pd.DataFrame({
    "Purchase Count":Pur_c,
    "Average Purchase Price":Pur_av,
    "Total Purchase Value":Pur_sum,
    "Avg Total Purchase per Person":Pur_C
})
Pur_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Demographic                                                               
<10                          32                  $3.40              $108.96   
10-14                        54                  $2.90              $156.60   
15-19                       101                  $3.04              $307.24   
20-24                       424                  $3.06             $1295.96   
25-29                        77                  $2.88              $221.42   
30-34                        52                  $2.99              $155.71   
35-39                        27                  $3.54               $95.64   
40+                          13                  $2.94               $38.24   

                Avg Total Purchase per Person  
Age Demographic                                
<10                                     $3.42  
10-14                                   $2.94  
15-19                                   $3.14  
20-24                                   $3.06  
25-29                                   $2.90  
30-34                                   $3.00  
35-39                                   $3.36  
40+                                     $3.04

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [23]:
Top_Spenders=purchase_data.groupby("SN")
TSP_Count=Top_Spenders["Price"].count()

In [24]:
TSP_Mean=['${:.2f}'.format(x) for x in Top_Spenders["Price"].mean()]
TSP_Mean

['$2.28',
 '$4.48',
 '$4.91',
 '$4.32',
 '$1.79',
 '$3.64',
 '$1.09',
 '$2.99',
 '$3.39',
 '$3.74',
 '$3.08',
 '$2.82',
 '$2.96',
 '$4.40',
 '$4.00',
 '$4.80',
 '$4.09',
 '$2.96',
 '$1.61',
 '$3.33',
 '$4.32',
 '$4.40',
 '$4.91',
 '$1.56',
 '$2.18',
 '$1.76',
 '$1.96',
 '$2.69',
 '$2.21',
 '$1.03',
 '$3.54',
 '$4.60',
 '$1.01',
 '$3.08',
 '$3.93',
 '$3.45',
 '$4.23',
 '$3.79',
 '$4.48',
 '$1.10',
 '$2.89',
 '$3.07',
 '$3.09',
 '$3.45',
 '$1.14',
 '$1.75',
 '$1.79',
 '$3.39',
 '$4.84',
 '$2.49',
 '$3.45',
 '$3.27',
 '$2.15',
 '$2.50',
 '$1.00',
 '$2.85',
 '$2.79',
 '$1.55',
 '$4.54',
 '$1.70',
 '$3.39',
 '$3.31',
 '$4.48',
 '$1.75',
 '$1.54',
 '$4.90',
 '$1.70',
 '$3.19',
 '$2.96',
 '$4.35',
 '$2.96',
 '$1.70',
 '$1.33',
 '$4.18',
 '$3.09',
 '$2.48',
 '$4.74',
 '$4.51',
 '$3.06',
 '$2.59',
 '$4.58',
 '$3.02',
 '$2.15',
 '$2.05',
 '$4.17',
 '$3.94',
 '$4.93',
 '$3.87',
 '$3.82',
 '$1.98',
 '$3.19',
 '$3.14',
 '$3.07',
 '$3.92',
 '$4.64',
 '$3.27',
 '$2.05',
 '$2.05',
 '$4.28',
 '$4.40',


In [25]:
TSP_Sum=['${:.2f}'.format(x) for x in Top_Spenders["Price"].sum()]
TSP_Sum

['$2.28',
 '$4.48',
 '$4.91',
 '$4.32',
 '$1.79',
 '$7.29',
 '$1.09',
 '$8.98',
 '$6.79',
 '$3.74',
 '$3.08',
 '$5.63',
 '$2.96',
 '$4.40',
 '$4.00',
 '$4.80',
 '$4.09',
 '$2.96',
 '$1.61',
 '$3.33',
 '$8.64',
 '$4.40',
 '$4.91',
 '$1.56',
 '$2.18',
 '$1.76',
 '$3.91',
 '$5.38',
 '$2.21',
 '$1.03',
 '$3.54',
 '$4.60',
 '$1.01',
 '$3.08',
 '$7.86',
 '$3.45',
 '$4.23',
 '$7.58',
 '$4.48',
 '$1.10',
 '$2.89',
 '$9.22',
 '$3.09',
 '$3.45',
 '$1.14',
 '$3.51',
 '$1.79',
 '$6.78',
 '$4.84',
 '$2.49',
 '$3.45',
 '$3.27',
 '$4.31',
 '$2.50',
 '$1.00',
 '$5.70',
 '$5.58',
 '$1.55',
 '$9.09',
 '$1.70',
 '$3.39',
 '$6.63',
 '$4.48',
 '$1.75',
 '$1.54',
 '$4.90',
 '$1.70',
 '$3.19',
 '$5.92',
 '$4.35',
 '$2.96',
 '$1.70',
 '$1.33',
 '$4.18',
 '$3.09',
 '$7.44',
 '$4.74',
 '$9.03',
 '$6.12',
 '$5.18',
 '$4.58',
 '$6.03',
 '$4.29',
 '$2.05',
 '$8.35',
 '$3.94',
 '$4.93',
 '$3.87',
 '$11.46',
 '$1.98',
 '$3.19',
 '$3.14',
 '$3.07',
 '$3.92',
 '$4.64',
 '$3.27',
 '$2.05',
 '$2.05',
 '$8.55',
 '$4.40',

In [26]:
TSP_df=pd.DataFrame({
    "Purchase Count":TSP_Count,
    "Average Purchase Price":TSP_Mean,
    "Total Purchase Value":TSP_Sum   
})
TSP_df

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                       
Adairialis76                1                  $2.28                $2.28
Adastirin33                 1                  $4.48                $4.48
Aeda94                      1                  $4.91                $4.91
Aela59                      1                  $4.32                $4.32
Aelaria33                   1                  $1.79                $1.79
...                       ...                    ...                  ...
Yathecal82                  3                  $2.07                $6.22
Yathedeu43                  2                  $3.01                $6.02
Yoishirrala98               1                  $4.58                $4.58
Zhisrisu83                  2                  $3.94                $7.89
Zontibe81                   3                  $2.68                $8.03

[576 rows x 3 columns]

In [27]:
A_TSP_df=TSP_df.sort_values(by="Total Purchase Value",ascending=False)
A_TSP_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                      
Haillyrgue51               3                  $3.17                $9.50
Phistym51                  2                  $4.75                $9.50
Lamil79                    2                  $4.64                $9.29
Aina42                     3                  $3.07                $9.22
Saesrideu94                2                  $4.59                $9.18

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [28]:
Pop_Items=purchase_data.groupby(["Item ID","Item Name"])

In [29]:
Pop_count=Pop_Items["Price"].count()
Pop_count

Item ID  Item Name                                   
0        Splinter                                         4
1        Crucifer                                         4
2        Verdict                                          6
3        Phantomlight                                     6
4        Bloodlord's Fetish                               5
                                                         ..
178      Oathbreaker, Last Hope of the Breaking Storm    12
179      Wolf, Promise of the Moonwalker                  6
181      Reaper's Toll                                    5
182      Toothpick                                        3
183      Dragon's Greatsword                              3
Name: Price, Length: 179, dtype: int64

In [30]:
Pop_TP=Pop_Items["Price"].sum()
Pop_TP

Item ID  Item Name                                   
0        Splinter                                         5.12
1        Crucifer                                        11.77
2        Verdict                                         14.88
3        Phantomlight                                    14.94
4        Bloodlord's Fetish                               8.50
                                                         ...  
178      Oathbreaker, Last Hope of the Breaking Storm    50.76
179      Wolf, Promise of the Moonwalker                 26.88
181      Reaper's Toll                                    8.30
182      Toothpick                                       12.09
183      Dragon's Greatsword                              3.27
Name: Price, Length: 179, dtype: float64

In [31]:
Item_price=Pop_Items["Price"].unique()
Item_price

Item ID  Item Name                                   
0        Splinter                                              [1.28]
1        Crucifer                                        [3.26, 1.99]
2        Verdict                                               [2.48]
3        Phantomlight                                          [2.49]
4        Bloodlord's Fetish                                     [1.7]
                                                             ...     
178      Oathbreaker, Last Hope of the Breaking Storm          [4.23]
179      Wolf, Promise of the Moonwalker                       [4.48]
181      Reaper's Toll                                         [1.66]
182      Toothpick                                             [4.03]
183      Dragon's Greatsword                                   [1.09]
Name: Price, Length: 179, dtype: object

In [32]:
Pop_df=pd.DataFrame({
    "Purchase Count":Pop_count,
    "Item Price":Item_price,
    "Total Purchase Value":['${:.2f}'.format(x) for x in Pop_TP]
})
Pop_df

Purchase Count  \
Item ID Item Name                                                      
0       Splinter                                                   4   
1       Crucifer                                                   4   
2       Verdict                                                    6   
3       Phantomlight                                               6   
4       Bloodlord's Fetish                                         5   
...                                                              ...   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
179     Wolf, Promise of the Moonwalker                            6   
181     Reaper's Toll                                              5   
182     Toothpick                                                  3   
183     Dragon's Greatsword                                        3   

                                                        Item Price  \
Item ID Item Name                                                    
0       Splinter                                            [1.28]   
1       Crucifer                                      [3.26, 1.99]   
2       Verdict                                             [2.48]   
3       Phantomlight                                        [2.49]   
4       Bloodlord's Fetish                                   [1.7]   
...                                                            ...   
178     Oathbreaker, Last Hope of the Breaking Storm        [4.23]   
179     Wolf, Promise of the Moonwalker                     [4.48]   
181     Reaper's Toll                                       [1.66]   
182     Toothpick                                           [4.03]   
183     Dragon's Greatsword                                 [1.09]   

                                                     Total Purchase Value  
Item ID Item Name                                                          
0       Splinter                                                    $5.12  
1       Crucifer                                                   $11.77  
2       Verdict                                                    $14.88  
3       Phantomlight                                               $14.94  
4       Bloodlord's Fetish                                          $8.50  
...                                                                   ...  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
179     Wolf, Promise of the Moonwalker                            $26.88  
181     Reaper's Toll                                               $8.30  
182     Toothpick                                                  $12.09  
183     Dragon's Greatsword                                         $3.27  

[179 rows x 3 columns]

In [33]:
A_Pop_df=Pop_df.sort_values(by="Purchase Count",ascending=False)
A_Pop_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                        Item Price  \
Item ID Item Name                                                    
92      Final Critic                                  [4.88, 4.19]   
178     Oathbreaker, Last Hope of the Breaking Storm        [4.23]   
145     Fiery Glass Crusader                                [4.58]   
132     Persuasion                                    [3.19, 3.33]   
108     Extraction, Quickblade Of Trembling Hands           [3.53]   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [34]:
Pop_df=pd.DataFrame({
    "Purchase Count":Pop_count,
    "Item Price":Item_price,
    "Total Purchase Value":Pop_TP
})
Pop_df.head()

Purchase Count    Item Price  Total Purchase Value
Item ID Item Name                                                             
0       Splinter                         4        [1.28]                  5.12
1       Crucifer                         4  [3.26, 1.99]                 11.77
2       Verdict                          6        [2.48]                 14.88
3       Phantomlight                     6        [2.49]                 14.94
4       Bloodlord's Fetish               5         [1.7]                  8.50

In [35]:
D_Pop_df=Pop_df.sort_values(by="Total Purchase Value",ascending=False)
D_Pop_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                        Item Price  \
Item ID Item Name                                                    
92      Final Critic                                  [4.88, 4.19]   
178     Oathbreaker, Last Hope of the Breaking Storm        [4.23]   
82      Nirvana                                              [4.9]   
145     Fiery Glass Crusader                                [4.58]   
103     Singed Scalpel                                      [4.35]   

                                                      Total Purchase Value  
Item ID Item Name                                                           
92      Final Critic                                                 59.99  
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
82      Nirvana                                                      44.10  
145     Fiery Glass Crusader                                         41.22  
103     Singed Scalpel                                               34.80

# Pandas Homework - Observable trends

### Gender Observation
* Males represents the 84% of players in the game
  * They have bought around $1968.00 in te period this mean more than the 75% purchases have been made by men.
  * Although per purchase women spent more than men.

### Purchases Observations
*The most profitable market for the videogame has between 20-24 years
    *Have purchase $1296.00

### Items Observations
*The most profitable item is the "Final Critic" with a total purchase value of $59.99 and a repetion of 13 purchases.